desc/compute/_fast_ion.py

```diff
@@ -354,3 +359,354 @@ def fun(pitch_inv):
     return data
+
+
+@register_compute_fun(
+    name="Jpar",
+    label=(
+        # J_∥ = ∫ dl v_∥ /(√E ∫ dl)
+        "\\J_{\\parallel} = \\integrate v_{\\parallel} dl/\\integrate dl/B"
+    ),
+    units="~",
+    units_long="~",
+    description="Normalized second adiabatic invariant of motion.",
+    coordinates="r",
+    dim=1,
+    profiles=[],
+    params=[],
+    transforms={"grid": []},
+    data=[
+        "min_tz |B|",
+        "max_tz |B|",
+        "B^zeta",
+    ]
+    + Bounce2D.required_names,
+    resolution_requirement="tz",
+    grid_requirement={"can_fft2": True},
+    **_bounce_doc,
+)
+@partial(
+    jit,
+    static_argnames=[
+        "Y_B",
+        "num_transit",
+        "num_well",
+        "num_quad",
+        "num_pitch",
+        "pitch_batch_size",
+        "surf_batch_size",
+        "spline",
+    ],
+)
+def _Jpar(params, transforms, profiles, data, **kwargs):
+    """Second adiabatic invariant of particle motion.
+
+    The normalization requires a length for which we have uesd the fieldline
+    length ∫ dl/B.
+    Typically calculated as a function of (rho, alpha, lambda)
+    """
+    # noqa: unused dependency
+    theta = kwargs["theta"]
+    Y_B = kwargs.get("Y_B", theta.shape[-1] * 2)
+    alpha = kwargs.get("alpha", jnp.array([0.0]))
+    num_transit = kwargs.get("num_transit", 20)
+    num_pitch = kwargs.get("num_pitch", 64)
+    num_well = kwargs.get("num_well", Y_B * num_transit)
+    pitch_batch_size = kwargs.get("pitch_batch_size", None)
+    surf_batch_size = kwargs.get("surf_batch_size", 1)
+    assert (
+        surf_batch_size == 1 or pitch_batch_size is None
+    ), f"Expected pitch_batch_size to be None, got {pitch_batch_size}."
+    spline = kwargs.get("spline", True)
+    fl_quad = (
+        kwargs["fieldline_quad"] if "fieldline_quad" in kwargs else leggauss(Y_B // 2)
+    )
+    quad = (
+        kwargs["quad"]
+        if "quad" in kwargs
+        else get_quadrature(
+            leggauss(kwargs.get("num_quad", 32)),
+            (automorphism_sin, grad_automorphism_sin),
+        )
+    )
+
+    def Jpar0(data):
+        bounce = Bounce2D(
+            grid,
+            data,
+            data["theta"],
+            Y_B,
+            alpha,
+            num_transit,
+            quad,
+            is_fourier=True,
+            spline=spline,
+        )
+
+        def fun(pitch_inv):
+            Jpar = bounce.integrate(
+                [_Jpar_num],
+                pitch_inv,
+                data,
+                ["B^zeta"],
+                bounce.points(pitch_inv, num_well),
+                is_fourier=True,
+            )
+            # Take sum over wells, mean in alpha
+            return jnp.sum(Jpar, axis=-1)
+
+        return batch_map(fun, data["pitch_inv"], pitch_batch_size) / (
+            bounce.compute_fieldline_length(fl_quad)
+        )
+
+    grid = transforms["grid"]
+    data["Jpar"] = _compute(
+        Jpar0,
+        {
+            "B^zeta": data["B^zeta"],
```

Comment:
@unalmis  This becomes a problem that causes unit test 3.12 to fail.
Basically, B^zeta is defined twice. I think the solution is straightforward for you to figure out but take me a couple of hours.

Reply:
> Basically, B^zeta is defined twice.

In the dependencies for the compute function, there was `B^zeta` + Bounce2D.requirednames = ["B^zeta", "B^zeta", ...]. The test was failing to warn of this.